diff -Naur linux-3.9.2/arch/mips/mm/c-r4k.c linux-3.9.2-PersianPrince/arch/mips/mm/c-r4k.c
--- linux-3.9.2/arch/mips/mm/c-r4k.c	2013-05-11 18:49:28.000000000 +0430
+++ linux-3.9.2-PersianPrince/arch/mips/mm/c-r4k.c	2021-03-24 23:35:30.852875397 +0430
@@ -632,6 +632,8 @@
 		if (size >= scache_size)
 			r4k_blast_scache();
 		else {
+			unsigned long lsize = cpu_scache_line_size();
+			unsigned long almask = ~(lsize - 1);
 			/*
 			 * There is no clearly documented alignment requirement
 			 * for the cache instruction on MIPS processors and
@@ -640,6 +642,9 @@
 			 * hit ops with insufficient alignment.	 Solved by
 			 * aligning the address to cache line size.
 			 */
+			cache_op(Hit_Writeback_Inv_SD, addr & almask);
+			cache_op(Hit_Writeback_Inv_SD,
+				 (addr + size - 1) & almask);
 			blast_inv_scache_range(addr, addr + size);
 		}
 		__sync();
@@ -649,7 +654,11 @@
 	if (cpu_has_safe_index_cacheops && size >= dcache_size) {
 		r4k_blast_dcache();
 	} else {
+		unsigned long lsize = cpu_dcache_line_size();
+		unsigned long almask = ~(lsize - 1);
 		R4600_HIT_CACHEOP_WAR_IMPL;
+		cache_op(Hit_Writeback_Inv_D, addr & almask);
+		cache_op(Hit_Writeback_Inv_D, (addr + size - 1)  & almask);
 		blast_inv_dcache_range(addr, addr + size);
 	}
 
